File: AddTwoNumbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        if(l1 == NULL){
            return l2;
        }
        if(l2 == NULL){
            return l1;
        }
        int carryBit = (l1->val + l2->val) / 10;
        ListNode* head = new ListNode((l1->val + l2->val) % 10);
        ListNode* prePivot = head;
        ListNode* pivot = NULL;
        ListNode* l1Pivot = l1->next;
        ListNode* l2Pivot = l2->next;
        while(l1Pivot != NULL && l2Pivot != NULL){
            int sum = l1Pivot->val + l2Pivot->val + carryBit;
            carryBit = sum / 10;
            pivot = new ListNode(sum % 10);
            prePivot->next = pivot;
            prePivot = pivot;
            l1Pivot = l1Pivot->next;
            l2Pivot = l2Pivot->next;
        }
        l1Pivot = l1Pivot == NULL ? l2Pivot : l1Pivot;
        while(l1Pivot != NULL){
            int sum = l1Pivot->val + carryBit;
            carryBit = sum / 10;
            pivot = new ListNode(sum % 10);
            prePivot->next = pivot;
            prePivot = pivot;
            l1Pivot = l1Pivot->next;
        }
        if(carryBit > 0){
            pivot = new ListNode(carryBit);
            prePivot->next = pivot;
        }
        return head;
        
    }
};

File: python_src/AddTwoNumbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = pivot = ListNode(0)
        carry_bit = 0
        l1_pivot = l1
        l2_pivot = l2
        while l1_pivot and l2_pivot:
            carry_bit, mod = divmod(carry_bit + l1_pivot.val + l2_pivot.val,10)
            pivot.next = ListNode(mod)
            pivot = pivot.next
            l1_pivot = l1_pivot.next
            l2_pivot = l2_pivot.next
            
        if l2_pivot: l1_pivot = l2_pivot
        while l1_pivot:
            carry_bit, mod = divmod(carry_bit + l1_pivot.val,10)
            pivot.next = ListNode(mod)
            pivot = pivot.next
            l1_pivot = l1_pivot.next
        if carry_bit != 0:
            pivot.next = ListNode(carry_bit)
        return head.next